src/hippopt/robot_planning/utilities/smooth_terrain.py

```diff
@@ -0,0 +1,475 @@
+import dataclasses
+from typing import TypeVar
+
+import casadi as cs
+import numpy as np
+
+from hippopt.robot_planning.utilities.terrain_descriptor import TerrainDescriptor
+from hippopt.robot_planning.utilities.terrain_sum import TerrainSum
+from hippopt.robot_planning.utilities.terrain_visualizer import (
+    TerrainVisualizer,
+    TerrainVisualizerSettings,
+)
+
+TSmoothTerrain = TypeVar("TSmoothTerrain", bound="SmoothTerrain")
+
+
+@dataclasses.dataclass
+class SmoothTerrain(TerrainDescriptor):
+    """
+    Smooth terrain is a terrain with a smooth height function.
+    The height is defined as follows:
+    h(x, y) = exp(−g(x, y)^(2s)) π(x, y).
+
+    Here, g(x, y) ≥ 0 is the equation of a closed curve in the xy-plane,
+    and π(x, y) is the equation of a plane in the xy-plane, defining the shape of the
+    terrain when exp(−g(x, y)^(2s)) = 1, i.e. when g(x, y) = 0.
+    The parameter s ≥ 1 controls the smoothness of the terrain.
```

Comment:
> Since for all (x, y) in the R^2, g(x, y) ≥ 0 by definition, it is no longer required for the exponent of g(x, y) to be an even number.

This is a good point. I wrote in the documentation that g(x, y) ≥ 0, but I have no way to check it, while I can control the exponent. So I can simply remove the constraint on g(x, y) to be always non-negative.



> Also, it is not necessary for the implicit equations g(x, y) = 0 or g(x, y) = 1 to represent closed curves in xy-plane. For istance, consider the g(x, y) = (x + y)^2 ≥ 0 (see image below):
> 
> * g(x, y) = 0 represents the line y = -x (defining the shape of the upper part of the terrain)
> * g(x, y) = 1 represents the two lines y = -x + 1 and y = -x - 1 (defining the shape of the lower part of the terrain)

Thanks! I will remove the term "closed"

Reply:
Done in https://github.com/ami-iit/hippopt/pull/10/commits/df7a3faaeb6bd69d0434984f8f83a8dda89d82de